pkg/autoscaler/autoscaler_server.go

```diff
@@ -0,0 +1,134 @@
+
+import (
+	"fmt"
+	"time"
+
+	"k8s.io/client-go/1.4/pkg/util/clock"
+
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/cmd/cluster-proportional-autoscaler/options"
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/controller"
+	"github.com/kubernetes-incubator/cluster-proportional-autoscaler/pkg/autoscaler/controller/laddercontroller"
+
+	"github.com/golang/glog"
+)
+
+// AutoScaler determines the number of replicas to run
+type AutoScaler struct {
+	K8sClient    K8sClientInterface
+	Controller   controller.Interface
+	ConfigMapKey string
+	PollPeriod   time.Duration
+	Clock        clock.Clock
+	StopCh       chan struct{}
+	ReadyCh      chan<- struct{} // For testing.
+}
+
+func NewAutoScaler(c *options.AutoScalerConfig) (*AutoScaler, error) {
+	var controller controller.Interface
+	switch c.Mode {
+	case laddercontroller.ControllerType:
+		controller = laddercontroller.NewLadderController()
+	default:
+		return nil, fmt.Errorf("not a supported control mode: %v", c.Mode)
+	}
+	newK8sClient, err := NewKubernetesClient(c.Namespace, c.Target)
+	if err != nil {
+		return nil, err
+	}
+	return &AutoScaler{
+		K8sClient:    newK8sClient,
+		Controller:   controller,
+		ConfigMapKey: c.ConfigMap,
+		PollPeriod:   time.Second * time.Duration(c.PollPeriodSeconds),
+		Clock:        clock.RealClock{},
+		StopCh:       make(chan struct{}),
+		ReadyCh:      make(chan struct{}, 1),
+	}, nil
+}
+
+// Run periodically counts the number of nodes and cores, estimates the expected
+// number of replicas, compares them to the actual replicas, and
+// updates the target resource with the expected replicas if necessary.
+func (s *AutoScaler) Run() {
+	firstTimeCh := make(chan struct{}, 1)
+	firstTimeCh <- struct{}{}
+	ticker := s.Clock.Tick(s.PollPeriod)
+	s.ReadyCh <- struct{}{} // For testing.
+
+	for {
+		select {
+		case <-firstTimeCh:
+			s.pollAPIServer()
+		case <-ticker:
+			s.pollAPIServer()
+		case <-s.StopCh:
+			return
+		}
+	}
+}
+
+func (s *AutoScaler) pollAPIServer() {
+	// Query the apiserver for the cluster status --- number of nodes and cores
+	clusterStatus, err := s.K8sClient.GetClusterStatus()
+	if err != nil {
+		glog.Errorf("error while counting nodes: %v\n", err)
+		return
+	}
+	glog.V(4).Infof("Total nodes %5d, schedulable nodes: %5d\n", clusterStatus.totalNodes, clusterStatus.schedulableNodes)
+	glog.V(4).Infof("Total cores %5d, schedulable cores: %5d\n", clusterStatus.totalCores, clusterStatus.schedulableCores)
+
+	// Fetch autoscaler ConfigMap data from apiserver
+	configMap, err := (s.K8sClient.FetchConfigMap(s.K8sClient.GetNamespace(), s.ConfigMapKey))
+	if err != nil {
+		glog.Errorf("error fetching scaler params: %s", err)
+		return
+	}
+
+	// Query the autoscaler for the expected replicas number
+	expReplicas, err := s.Controller.GetExpectedNumOfReplicas(configMap.Data[laddercontroller.ControllerType], configMap.Version, clusterStatus.schedulableNodes, clusterStatus.schedulableCores)
```

Comment:
Would this be cleaner if you broke it up:

```
// Update the config
s.Controller.SyncConfig(configMap)

// Figure out what we have
expReplicas, err := s.Controller.GetExpectedNumOfReplicas(clusterStatus.schedulableNodes, clusterStatus.schedulableCores)
```
￼
or even:

```
expReplicas, err := s.Controller.GetExpectedNumOfReplicas(configMap, clusterStatus.schedulableNodes, clusterStatus.schedulableCores)
```

In theory you can have other types of controllers (not ladder) but here you used the ladder type without indirecting.  It seems simpler to just pass the whole config and let the controller extract what it needs.